File: mcp/src/blender/tools/core/file.js
// File tools — blender_file_info, blender_file_save, blender_file_open

import { sendToBlender } from "../../blender-bridge.js";

export const fileTools = [
  {
    name: "blender_file_info",
    description:
      "Get information about the current Blender file: filepath, save status, " +
      "Blender version, active scene, and object/mesh/material counts. " +
      "Returns: { filepath, is_saved, is_dirty, blender_version, active_scene, object_count, mesh_count, material_count }",
    inputSchema: {
      type: "object",
      properties: {},
    },
    handler: async () => {
      const result = await sendToBlender("file/info", {});
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_file_save",
    description:
      "Save the current Blender file. " +
      "If filepath is provided the file is saved to that path (Save As). " +
      "If no filepath is provided the file is saved in-place; the file must have been " +
      "saved at least once before, otherwise an error is returned. " +
      "Example call: { \"filepath\": \"C:/Projects/scene_v2.blend\", \"compress\": true }. " +
      "Returns: { success, filepath, compressed }",
    inputSchema: {
      type: "object",
      properties: {
        filepath: {
          type: "string",
          description:
            "Destination path for the .blend file. " +
            "Example: \"C:/Projects/scene.blend\". Omit to save in-place.",
        },
        compress: {
          type: "boolean",
          description:
            "Write a compressed .blend file when true. " +
            "IMPORTANT: Must be a JSON boolean (true or false), NOT a string. " +
            "Defaults to false.",
        },
      },
    },
    handler: async (args) => {
      const result = await sendToBlender("file/save", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_file_open",
    description:
      "Open a .blend file in Blender. " +
      "WARNING: This reloads the entire Blender session — all unsaved changes will be lost. " +
      "Example call: { \"filepath\": \"C:/Projects/scene.blend\" }. " +
      "Returns: { success, filepath, active_scene, object_count }",
    inputSchema: {
      type: "object",
      properties: {
        filepath: {
          type: "string",
          description: "Absolute path to the .blend file to open. Example: \"C:/Projects/scene.blend\".",
        },
      },
      required: ["filepath"],
    },
    handler: async (args) => {
      const result = await sendToBlender("file/open", args);
      return JSON.stringify(result, null, 2);
    },
  },
];

File: src/scripts/mixar/modules/mcp_bridge/blender/handlers/file.py
"""
File handlers for Blender MCP Bridge.
Provides file-related operations: info, save, open.
"""

import bpy
from ..utils.response import ok_response, error_response, validate_filepath
from ..utils.context_helpers import safe_operator_call
from . import register_handler


def _file_info(params):
    """
    Get information about the current Blender file.

    Route: POST /api/file/info

    Returns:
        filepath, is_saved, is_dirty, blender_version, active_scene
    """
    try:
        filepath = bpy.data.filepath
        return ok_response({
            "filepath": filepath if filepath else "",
            "is_saved": bool(filepath),
            "is_dirty": bpy.data.is_dirty,
            "blender_version": ".".join(str(v) for v in bpy.app.version),
            "active_scene": bpy.context.scene.name,
            "object_count": len(bpy.data.objects),
            "mesh_count": len(bpy.data.meshes),
            "material_count": len(bpy.data.materials),
        })
    except Exception as e:
        return error_response(f"Failed to get file info: {e}")


def _file_save(params):
    """
    Save the current Blender file.

    Route: POST /api/file/save

    Optional params:
        filepath (str): Destination path (Save As). Omit to save in-place.
        compress (bool): Write a compressed .blend file. Defaults to False.

    Returns:
        {filepath, compressed}
    """
    try:
        filepath = params.get("filepath", "").strip()
        compress = bool(params.get("compress", False))

        if filepath:
            filepath, path_err = validate_filepath(filepath)
            if path_err:
                return error_response(path_err)

        if filepath:
            success, result = safe_operator_call(
                bpy.ops.wm.save_as_mainfile,
                filepath=filepath,
                compress=compress,
            )
            if not success:
                return error_response(f"Save As failed: {result}")
            saved_path = filepath
        else:
            current_path = bpy.data.filepath
            if not current_path:
                return error_response(
                    "The file has never been saved. "
                    "Provide a 'filepath' parameter to save it for the first time."
                )
            success, result = safe_operator_call(
                bpy.ops.wm.save_mainfile,
                compress=compress,
            )
            if not success:
                return error_response(f"Save failed: {result}")
            saved_path = current_path

        return ok_response({"filepath": saved_path, "compressed": compress})
    except Exception as e:
        return error_response(f"Failed to save file: {e}")


def _file_open(params):
    """
    Open a .blend file in Blender.

    Route: POST /api/file/open

    WARNING: This reloads the entire Blender session — unsaved changes are lost.

    Required params:
        filepath (str): Absolute path to the .blend file to open.

    Returns:
        {filepath, active_scene, object_count}
    """
    filepath = params.get("filepath", "").strip()
    if not filepath:
        return error_response("Parameter 'filepath' is required.")

    filepath, path_err = validate_filepath(filepath, must_exist=True)
    if path_err:
        return error_response(path_err)

    try:
        success, result = safe_operator_call(
            bpy.ops.wm.open_mainfile,
            filepath=filepath,
        )
        if not success:
            return error_response(f"Open failed: {result}")

        return ok_response({
            "filepath": filepath,
            "active_scene": bpy.context.scene.name,
            "object_count": len(bpy.data.objects),
        })
    except Exception as e:
        return error_response(f"Failed to open file: {e}")


# ─── Register routes ───
register_handler("file", "info", _file_info)
register_handler("file", "save", _file_save)
register_handler("file", "open", _file_open)
